### PR TITLE
CFE-3205: Added acceptance test to check syntax of examples

### PR DESCRIPTION
**Pull request by olehermanse**
Runs cf-promises on all examples.

### CODE REVIEW

File: tests/acceptance/04_examples/check_syntax.cf [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+#######################################################
+
+bundle agent test
+{
+  meta:
+      # It's unrealistic to keep this test perfectly stable on all non-Linux platforms.
+      # It would require examples that quickly get very cluttered.
+      "test_skip_unsupported" string => "!linux";
```
Comment (nickanderson) [inline tests/acceptance/04_examples/check_syntax.cf]: Please add a `"description" string => "Syntax check each policy (.cf) file in examples";` meta var.

File: tests/acceptance/04_examples/check_syntax.cf [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    "Ran checker '$(checker) $(file)'"
+      handle => "printed_command_$(cfile)";
+    "Syntax error in example: '$(file)'"
+      depends_on => { "printed_command_$(cfile)" },
+      if => "failure_$(cfile)",
+      handle => "reported_error_$(cfile)";
+
+  methods:
+    "verbose_$(cfile)"
+      usebundle => verbose_output("$(checker) $(file)"),
```
Comment (nickanderson) [inline tests/acceptance/04_examples/check_syntax.cf]: So, you run the checker again, with verbose output so that you can display the specific error?

Comment (nickanderson) [reply tests/acceptance/04_examples/check_syntax.cf]: No, not run with verbose output, just run as a commands promise.

File: tests/acceptance/04_examples/check_syntax.cf [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  commands:
+      "$(runme)";
+}
+
+#######################################################
+
+bundle agent check
+{
+  classes:
+      "ok" not => classmatch("failure.*");
```
Comment (nickanderson) [inline tests/acceptance/04_examples/check_syntax.cf]: I would `classmatch("failure_.*")`.

File: tests/acceptance/04_examples/check_syntax.cf [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+  reports:
+    ok::
+      "$(this.promise_filename) Pass";
+    !ok::
```
Comment (nickanderson) [inline tests/acceptance/04_examples/check_syntax.cf]: Might consider adding a new bundle to dcs.cf.sub for pass/fail based on classmatch regex.

https://github.com/cfengine/core/blob/master/tests/acceptance/dcs.cf.sub#L747

File: tests/acceptance/04_examples/check_syntax.cf (hunk 1 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/check_syntax.cf]: Added.

File: tests/acceptance/04_examples/check_syntax.cf (hunk 2 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/check_syntax.cf]: Note that many things are copied from the check_outputs test. I changed the name here to `run_with_output`.

File: tests/acceptance/04_examples/check_syntax.cf (hunk 4 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/check_syntax.cf]: Ok, but outside of scope for now.

File: tests/acceptance/04_examples/check_syntax.cf (hunk 3 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/check_syntax.cf]: Added.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4216/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4216/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4216/

File: examples/neighbourhood_watch.cf [hunk 5]
```diff
@@ -1,4 +1,4 @@
-a#  Copyright 2019 Northern.tech AS
+#  Copyright 2019 Northern.tech AS
```
Comment (craigcomstock) [inline examples/neighbourhood_watch.cf]: nice! :p

File: tests/acceptance/04_examples/syntax/check_syntax.cf [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        not => returnszero("$(checker) $(file)", "noshell"),
+        scope => "namespace";
+
+  reports:
+    DEBUG::
+    # "Ran checker '$(checker) $(file)'"
+    #   handle => "printed_command_$(cfile)";
+    "Syntax error in example: '$(file)'"
+      # depends_on => { "printed_command_$(cfile)" },
+      if => "failure_$(cfile)",
```
Comment (craigcomstock) [inline tests/acceptance/04_examples/syntax/check_syntax.cf]: remove these commented out bits?

Comment (olehermanse) [reply tests/acceptance/04_examples/syntax/check_syntax.cf]: Yep, thanks, fixed.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4217/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4217/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4217/

File: tests/acceptance/04_examples/syntax/check_syntax.cf [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+# Symlink a file, only if dst does not exist
+# If dst already exists, assume it's correct (testing with masterfiles)
+{
+  classes:
+    "file_missing"
+      if => not(fileexists("$(dst)"));
+  methods:
+    file_missing::
+      "symlink"
+        usebundle => symlink("$(src)", "$(dst)");
```
Comment (vpodzime) [inline tests/acceptance/04_examples/syntax/check_syntax.cf]: You could just do `if => ...` on the `methods` promise, right?

File: tests/acceptance/04_examples/syntax/check_syntax.cf [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+#######################################################
+
+bundle agent test
+{
+  meta:
+      # It's unrealistic to keep this test perfectly stable on all non-Linux platforms.
+      # It would require examples that quickly get very cluttered.
+      "test_skip_unsupported"
+        string => "!linux";
```
Comment (vpodzime) [inline tests/acceptance/04_examples/syntax/check_syntax.cf]: Why should a syntax check of examples have problems on non-linux platforms?

File: tests/acceptance/04_examples/syntax/check_syntax.cf (hunk 7 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/syntax/check_syntax.cf]: Yes, but I think this is the more "normal" way to do it. Define classes to make decisions using class guards. `if` (previously `ifvarclass`) has, at least traditionally, 1 specific purpose: To be able to use variables in class expressions.

File: tests/acceptance/04_examples/syntax/check_syntax.cf (hunk 8 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/syntax/check_syntax.cf]: This part is copied from the other test. We could try without it.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, with exotics, please.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/4220/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/4220/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-4220/

File: tests/acceptance/04_examples/syntax/check_syntax.cf (hunk 8 above)
Comment (olehermanse) [reply tests/acceptance/04_examples/syntax/check_syntax.cf]: Let's do this in a follow up, when we have exotics available.

**Pull request merged**